File: backend/models/engagement.py

from sqlalchemy import Column, String, Integer, ForeignKey, DateTime, Enum, Text
from sqlalchemy.orm import relationship
from database.types import UUID
import uuid
from datetime import datetime
import enum

from database.connection import Base

class ContentType(str, enum.Enum):
    PLAN = "plan"
    CHAPTER = "chapter"
    QUIZ = "quiz"
    SIMULATION = "simulation"

class EngagementAction(str, enum.Enum):
    LIKE = "like"
    DISLIKE = "dislike"
    RATE = "rate"

class Engagement(Base):
    __tablename__ = "engagements"

    id = Column(UUID, primary_key=True, default=uuid.uuid4)
    user_id = Column(UUID, ForeignKey("users.id", ondelete="CASCADE"), nullable=False)
    
    content_type = Column(String(50), nullable=False) # Store as string for flexibility, validated by Pydantic
    content_id = Column(String(255), nullable=False)
    
    action = Column(String(20), nullable=False) # like, dislike, rate
    value = Column(Integer, nullable=False) # 1/-1 for like/dislike, 1-5 for rate
    comment = Column(Text, nullable=True)
    
    created_at = Column(DateTime, default=datetime.utcnow)

    # Relationships
    user = relationship("User", backref="engagements")

File: frontend/src/api/engagement.ts

import apiClient from './client';

export type ContentType = 'plan' | 'chapter' | 'quiz' | 'simulation';
export type EngagementAction = 'like' | 'dislike' | 'rate';

export interface EngagementData {
  content_type: ContentType;
  content_id: string;
  action: EngagementAction;
  value: number;
  comment?: string;
}

export interface EngagementResponse {
  id: string;
  content_type: ContentType;
  content_id: string;
  action: EngagementAction;
  value: number;
  comment?: string;
}

export const submitEngagement = async (data: EngagementData) => {
  const response = await apiClient.post<EngagementResponse>('/engagement', data);
  return response.data;
};

export const getEngagement = async (
  contentType: ContentType,
  contentId: string,
  action: EngagementAction
) => {
  const response = await apiClient.get<EngagementResponse | null>('/engagement/me', {
    params: {
      content_type: contentType,
      content_id: contentId,
      action
    }
  });
  return response.data;
};
